### PR TITLE
ROX-17300: Support hostAliases

### PR DESCRIPTION
**Pull request by ludydoo**
## Description

Adding `hostAliases` field (helm + CRD) for all components. 

Docs: https://github.com/openshift/openshift-docs/pull/76133

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] ~~Evaluated and added CHANGELOG entry if required~~
- [ ] ~~Determined and documented upgrade steps~~
- [x] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs)): https://github.com/openshift/openshift-docs/pull/76133

If any of these don't apply, please comment below.

### Here I tell how I validated my change

- Added unit tests for CR -> Helm values translation
- Added unit tests for helm rendering expectations (helmtest)
- Added e2e/kuttl tests for testing the operator behavior

Manual steps:
- Using both operator and helm installation methods, fill-in all `hostAliases` options
- Install the product using both methods
- Verify that the manifests include the expected hostAliases

Details below:

#### Specifying all hostAliases options for all components, and verifying that they are correctly applied by the operator: 

<details>
<summary>Install using the Operator</summary>

  ```bash
  #!/bin/bash

  cat <<EOF | kubectl apply -f -
  apiVersion: v1
  kind: Namespace
  metadata:
    name: stackrox-central
  ---
  apiVersion: v1
  kind: Namespace
  metadata:
    name: stackrox-secured-cluster
  EOF

  cd operator

  NAMESPACE=stackrox-central make stackrox-pull-secret
  NAMESPACE=stackrox-secured-cluster make stackrox-pull-secret

  cat <<EOF | kubectl apply -f -
  apiVersion: v1
  kind: Secret
  metadata:
    name: admin-pass
    namespace: stackrox-central
  stringData:
    password: letmein
  ---
  apiVersion: platform.stackrox.io/v1alpha1
  kind: Central
  metadata:
    name: central
    namespace: stackrox-central
  spec:
    imagePullSecrets:
      - name: stackrox
    monitoring:
      openshift:
        enabled: false
    central:
      adminPasswordSecret:
        name: admin-pass
      resources:
        requests:
          cpu: 10m
          memory: 50Mi
      hostAliases:
        - ip: "127.0.0.1"
          hostnames: ["central.com"]
      db:
        resources:
          requests:
            cpu: 10m
            memory: 50Mi
        hostAliases:
          - ip: "127.0.0.1"
            hostnames: ["central-db.com"]
    scanner:
      scannerComponent: Enabled
      analyzer:
        resources:
          requests:
            cpu: 10m
            memory: 50Mi
        hostAliases:
          - ip: "127.0.0.1"
            hostnames: ["scanner.com"]
      db:
        resources:
          requests:
            cpu: 10m
            memory: 50Mi
        hostAliases:
          - ip: "127.0.0.1"
            hostnames: ["scanner-db.com"]
    scannerV4:
      scannerComponent: Enabled
      matcher:
        resources:
          requests:
            cpu: 10m
            memory: 50Mi
        hostAliases:
          - ip: "127.0.0.1"
            hostnames: ["scanner-v4-matcher.com"]
      indexer:
        resources:
          requests:
            cpu: 10m
            memory: 50Mi
        hostAliases:
          - ip: "127.0.0.1"
            hostnames: ["scanner-v4-indexer.com"]
      db:
        resources:
          requests:
            cpu: 10m
            memory: 50Mi
        hostAliases:
          - ip: "127.0.0.1"
            hostnames: ["scanner-v4-db.com"]
  ---
  apiVersion: platform.stackrox.io/v1alpha1
  kind: SecuredCluster
  metadata:
    name: stackrox-secured-cluster-services
    namespace: stackrox-secured-cluster
  spec:
    clusterName: my-cluster
    imagePullSecrets:
      - name: stackrox
    sensor:
      resources:
          requests:
            cpu: 10m
            memory: 50Mi
      hostAliases:
        - ip: "127.0.0.1"
          hostnames: ["sensor.com"]
    admissionControl:
      resources:
          requests:
            cpu: 10m
            memory: 50Mi
      hostAliases:
        - ip: "127.0.0.1"
          hostnames: ["admission-control.com"]
    perNode:
      collector:
        resources:
          requests:
            cpu: 10m
            memory: 50Mi
      hostAliases:
        - ip: "127.0.0.1"
          hostnames: ["collector.com"]
    scanner:
      scannerComponent: AutoSense
      analyzer:
        resources:
          requests:
            cpu: 10m
            memory: 50Mi
        hostAliases:
          - ip: "127.0.0.1"
            hostnames: ["scanner.com"]
      db:
        resources:
          requests:
            cpu: 10m
            memory: 50Mi
        hostAliases:
          - ip: "127.0.0.1"
            hostnames: ["scanner-db.com"]
    scannerV4:
      scannerComponent: AutoSense
      indexer:
        resources:
          requests:
            cpu: 10m
            memory: 50Mi
        hostAliases:
          - ip: "127.0.0.1"
            hostnames: ["scanner-v4-indexer.com"]
      db:
        resources:
          requests:
            cpu: 10m
            memory: 50Mi
        hostAliases:
          - ip: "127.0.0.1"
            hostnames: ["scanner-v4-db.com"]
  ---
  EOF


  echo "Creating init bundle"
  kubectl -n stackrox-central exec deploy/central -- \
    roxctl central init-bundles generate init-bundle --insecure-skip-tls-verify --password letmein --output-secrets - \
    | kubectl -n stackrox-secured-cluster apply -f -


  echo "[Central] Getting central hostAliases"
  kubectl get deployment central -n stackrox-central -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

  echo "[Central] Getting scanner hostAliases"
  kubectl get deployment scanner -n stackrox-central -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

  echo "[Central] Getting scanner-db hostAliases"
  kubectl get deployment scanner-db -n stackrox-central -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

  echo "[Central] Getting scannerV4 indexer hostAliases"
  kubectl get deployment scanner-v4-indexer -n stackrox-central -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

  echo "[Central] Getting scannerV4 matcher hostAliases"
  kubectl get deployment scanner-v4-matcher -n stackrox-central -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

  echo "[Central] Getting scannerV4 db hostAliases"
  kubectl get deployment scanner-v4-db -n stackrox-central -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

  echo "[SecuredCluster] Getting sensor hostAliases"
  kubectl get deployment sensor -n stackrox-secured-cluster -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

  echo "[SecuredCluster] Getting admission-control hostAliases"
  kubectl get deployment admission-control -n stackrox-secured-cluster -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

  echo "[SecuredCluster] Getting collector hostAliases"
  kubectl get daemonset collector -n stackrox-secured-cluster -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

  echo "[SecuredCluster] Getting scanner hostAliases"
  kubectl get deployment scanner -n stackrox-secured-cluster -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

  echo "[SecuredCluster] Getting scanner-db hostAliases"
  kubectl get deployment scanner-db -n stackrox-secured-cluster -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

  echo "[SecuredCluster] Getting scannerV4 indexer hostAliases"
  kubectl get deployment scanner-v4-indexer -n stackrox-secured-cluster -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

  echo "[SecuredCluster] Getting scannerV4 db hostAliases"
  kubectl get deployment scanner-v4-db -n stackrox-secured-cluster -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'
  ```
</details>

Output
```
[Central] Getting central hostAliases
[{"hostnames":["central.com"],"ip":"127.0.0.1"}]

[Central] Getting scanner hostAliases
[{"hostnames":["scanner.com"],"ip":"127.0.0.1"}]

[Central] Getting scanner-db hostAliases
[{"hostnames":["scanner-db.com"],"ip":"127.0.0.1"}]

[Central] Getting scannerV4 indexer hostAliases
[{"hostnames":["scanner-v4-indexer.com"],"ip":"127.0.0.1"}]

[Central] Getting scannerV4 matcher hostAliases
[{"hostnames":["scanner-v4-matcher.com"],"ip":"127.0.0.1"}]

[Central] Getting scannerV4 db hostAliases
[{"hostnames":["scanner-v4-db.com"],"ip":"127.0.0.1"}]

[SecuredCluster] Getting sensor hostAliases
[{"hostnames":["sensor.com"],"ip":"127.0.0.1"}]

[SecuredCluster] Getting admission-control hostAliases
[{"hostnames":["admission-control.com"],"ip":"127.0.0.1"}]

[SecuredCluster] Getting collector hostAliases
[{"hostnames":["collector.com"],"ip":"127.0.0.1"}]

[SecuredCluster] Getting scanner hostAliases
[{"hostnames":["scanner.com"],"ip":"127.0.0.1"}]

[SecuredCluster] Getting scanner-db hostAliases
[{"hostnames":["scanner-db.com"],"ip":"127.0.0.1"}]

[SecuredCluster] Getting scannerV4 indexer hostAliases
[{"hostnames":["scanner-v4-indexer.com"],"ip":"127.0.0.1"}]

[SecuredCluster] Getting scannerV4 db hostAliases
[{"hostnames":["scanner-v4-db.com"],"ip":"127.0.0.1"}]
```

<details>
<summary>Install using helm</summary>

```bash
#!/bin/bash

roxctl helm output central-services
roxctl helm output secured-cluster-services

cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Namespace
metadata:
  name: stackrox-central
---
apiVersion: v1
kind: Namespace
metadata:
  name: stackrox-secured-cluster
---
apiVersion: v1
kind: Secret
metadata:
  name: admin-pass
  namespace: stackrox-central
stringData:
  password: letmein
EOF

cd operator

NAMESPACE=stackrox-central make stackrox-image-pull-secret
NAMESPACE=stackrox-secured-cluster make stackrox-image-pull-secret

cat <<EOF > values-central.yaml
allowNonstandardNamespace: true
imagePullSecrets:
  useExisting: 
    - stackrox
image:
  registry: quay.io/rhacs-eng
env:
  openshift: 4
central:
  adminPassword:
    value: letmein
  persistence:
    none: true
  image:
    tag: "4.4.x-722-g40ca8d1935"
  hostAliases:
    - ip: "127.0.0.1"
      hostnames: ["central.com"]
  db:
    image:
      tag: "4.4.x-722-g40ca8d1935"
    hostAliases:
      - ip: "127.0.0.1"
        hostnames: ["central-db.com"]
scanner:
  disable: false
  image:
    tag: "4.4.x-722-g40ca8d1935"
  dbImage:
    tag: "4.4.x-722-g40ca8d1935"
  hostAliases:
    - ip: "127.0.0.1"
      hostnames: ["scanner.com"]
  dbHostAliases:
    - ip: "127.0.0.1"
      hostnames: ["scanner-db.com"]
scannerV4:
  disable: false
  indexer:
    image:
      tag: "4.4.x-722-g40ca8d1935"
    hostAliases:
    - ip: "127.0.0.1"
      hostnames: ["scanner-v4-indexer.com"]
  matcher:
    image:
      tag: "4.4.x-722-g40ca8d1935"
    hostAliases:
    - ip: "127.0.0.1"
      hostnames: ["scanner-v4-matcher.com"]
  db:
    image:
      tag: "4.4.x-722-g40ca8d1935"
    hostAliases:
    - ip: "127.0.0.1"
      hostnames: ["scanner-v4-db.com"]
EOF

helm upgrade --install -n stackrox-central \
  -f values-central.yaml \
    stackrox-central-services ./stackrox-central-services-chart

echo "Creating init bundle"
kubectl port-forward -n stackrox-central svc/central 8443:443 &
PID=$!
roxctl central init-bundles generate init-bundle --insecure-skip-tls-verify --password letmein --output init-bundle.yaml -e localhost:8443
kill $PID

cat <<EOF > values-secured-cluster.yaml
allowNonstandardNamespace: true
clusterName: my-cluster
centralEndpoint: central.stackrox-central.svc:443
imagePullSecrets:
  useExisting: 
    - stackrox
env:
  openshift: 4
image:
  registry: quay.io/rhacs-eng
  main:
    tag: "4.4.x-722-g40ca8d1935"
  collector:
    tag: "4.4.x-722-g40ca8d1935"
  scanner:
    tag: "4.4.x-722-g40ca8d1935"
  scannerV4:
    tag: "4.4.x-722-g40ca8d1935"
  scannerV4DB:
    tag: "4.4.x-722-g40ca8d1935"
sensor:
  hostAliases:
    - ip: "127.0.0.1"
      hostnames: ["sensor.com"]
admissionControl:
  hostAliases:
    - ip: "127.0.0.1"
      hostnames: ["admission-control.com"]
collector:
  hostAliases:
    - ip: "127.0.0.1"
      hostnames: ["collector.com"]
scanner:
  disable: false
  hostAliases:
    - ip: "127.0.0.1"
      hostnames: ["scanner.com"]
  dbHostAliases:
    - ip: "127.0.0.1"
      hostnames: ["scanner-db.com"]
scannerV4:
  disable: false
  indexer:
    hostAliases:
    - ip: "127.0.0.1"
      hostnames: ["scanner-v4-indexer.com"]
  db:
    hostAliases:
    - ip: "127.0.0.1"
      hostnames: ["scanner-v4-db.com"]
EOF

helm upgrade --install -n stackrox-secured-cluster \
  -f values-secured-cluster.yaml \
  -f init-bundle.yaml \
    stackrox-secured-cluster-services ./stackrox-secured-cluster-services-chart

echo "[Central] Getting central hostAliases"
kubectl get deployment central -n stackrox-central -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

echo "[Central] Getting scanner hostAliases"
kubectl get deployment scanner -n stackrox-central -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

echo "[Central] Getting scanner-db hostAliases"
kubectl get deployment scanner-db -n stackrox-central -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

echo "[Central] Getting scannerV4 indexer hostAliases"
kubectl get deployment scanner-v4-indexer -n stackrox-central -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

echo "[Central] Getting scannerV4 matcher hostAliases"
kubectl get deployment scanner-v4-matcher -n stackrox-central -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

echo "[Central] Getting scannerV4 db hostAliases"
kubectl get deployment scanner-v4-db -n stackrox-central -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

echo "[SecuredCluster] Getting sensor hostAliases"
kubectl get deployment sensor -n stackrox-secured-cluster -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

echo "[SecuredCluster] Getting admission-control hostAliases"
kubectl get deployment admission-control -n stackrox-secured-cluster -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

echo "[SecuredCluster] Getting collector hostAliases"
kubectl get daemonset collector -n stackrox-secured-cluster -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

echo "[SecuredCluster] Getting scanner hostAliases"
kubectl get deployment scanner -n stackrox-secured-cluster -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

echo "[SecuredCluster] Getting scanner-db hostAliases"
kubectl get deployment scanner-db -n stackrox-secured-cluster -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

echo "[SecuredCluster] Getting scannerV4 indexer hostAliases"
kubectl get deployment scanner-v4-indexer -n stackrox-secured-cluster -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

echo "[SecuredCluster] Getting scannerV4 db hostAliases"
kubectl get deployment scanner-v4-db -n stackrox-secured-cluster -o jsonpath='{.spec.template.spec.hostAliases}{"\n\n"}'

```
</details>

#### Output
```
[Central] Getting central hostAliases
[{"hostnames":["central.com"],"ip":"127.0.0.1"}]

[Central] Getting scanner hostAliases
[{"hostnames":["scanner.com"],"ip":"127.0.0.1"}]

[Central] Getting scanner-db hostAliases
[{"hostnames":["scanner-db.com"],"ip":"127.0.0.1"}]

[Central] Getting scannerV4 indexer hostAliases
[{"hostnames":["scanner-v4-indexer.com"],"ip":"127.0.0.1"}]

[Central] Getting scannerV4 matcher hostAliases
[{"hostnames":["scanner-v4-matcher.com"],"ip":"127.0.0.1"}]

[Central] Getting scannerV4 db hostAliases
[{"hostnames":["scanner-v4-db.com"],"ip":"127.0.0.1"}]

[SecuredCluster] Getting sensor hostAliases
[{"hostnames":["sensor.com"],"ip":"127.0.0.1"}]

[SecuredCluster] Getting admission-control hostAliases
[{"hostnames":["admission-control.com"],"ip":"127.0.0.1"}]

[SecuredCluster] Getting collector hostAliases
[{"hostnames":["collector.com"],"ip":"127.0.0.1"}]

[SecuredCluster] Getting scanner hostAliases
[{"hostnames":["scanner.com"],"ip":"127.0.0.1"}]

[SecuredCluster] Getting scanner-db hostAliases
[{"hostnames":["scanner-db.com"],"ip":"127.0.0.1"}]

[SecuredCluster] Getting scannerV4 indexer hostAliases
[{"hostnames":["scanner-v4-indexer.com"],"ip":"127.0.0.1"}]

[SecuredCluster] Getting scannerV4 db hostAliases
[{"hostnames":["scanner-v4-db.com"],"ip":"127.0.0.1"}]
```

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: image/templates/helm/shared/templates/02-scanner-06-deployment.yaml.htpl [hunk 1]
```diff
@@ -41,6 +41,9 @@ spec:
       tolerations:
         {{- toYaml ._rox.scanner.tolerations | nindent 8 }}
       {{- end }}
+      {{- if ._rox.scanner.hostAliases }}
+      hostAliases: {{ toYaml ._rox.scanner.hostAliases | nindent 8 }}
+      {{- end }}
```
Comment (johannes94) [inline image/templates/helm/shared/templates/02-scanner-06-deployment.yaml.htpl:46]: I think you'll have to move that to the next line.
```suggestion
      {{- if ._rox.scanner.hostAliases }}
      hostAliases:
        {{ toYaml ._rox.scanner.hostAliases | nindent 8 }}
      {{- end }}
```

**Review by johannes94 — CHANGES_REQUESTED** [review]
Could you also add either a helmtest or a kuttl test that actually sets that setting on CRs / helm values to ensure rendered helm charts are valid?

File: image/templates/helm/shared/templates/02-scanner-06-deployment.yaml.htpl (hunk 1 above)
Comment (ludydoo) [reply image/templates/helm/shared/templates/02-scanner-06-deployment.yaml.htpl:46]: @johannes94 not with `nindent` (as opposed to `indent`) - it adds a newline

File: image/templates/helm/stackrox-secured-cluster/templates/admission-controller.yaml [hunk 2]
```diff
@@ -44,6 +44,9 @@ spec:
       {{- if ._rox.admissionControl.priorityClassName }}
       priorityClassName: {{ ._rox.admissionControl.priorityClassName }}
       {{- end }}
+      {{- if ._rox.admissionControl.hostAliases }}
+      hostAliases: {{ toYaml ._rox.admissionControl.hostAliases | nindent 8 }}
+      {{ - end }}
```
Comment (johannes94) [inline image/templates/helm/stackrox-secured-cluster/templates/admission-controller.yaml:49]:
```suggestion
      {{- end }}
```

File: image/templates/helm/shared/templates/02-scanner-v4-07-db-deployment.yaml [hunk 3]
```diff
@@ -41,6 +41,9 @@ spec:
       {{- end }}
       affinity:
         {{- toYaml ._rox.scannerV4.db.affinity | nindent 8 }}
+      {{- if ._rox.scannerV4.db.hostAliases }}
+      hostAliases: {{ toYaml ._rox.scannerV4.db.hostAliases | nindent 8 }}
+      {{ - end }}
```
Comment (johannes94) [inline image/templates/helm/shared/templates/02-scanner-v4-07-db-deployment.yaml:46]:
```suggestion
      {{- end }}
```

**Review by johannes94 — APPROVED** [review]
Looks good, just a minor thing about the field comments.

**Comment by abdennour** [comment]
Thanks for listing to our request

**Pull request merged**